### ISSUE TITLE
Dataflow + Starboard?

### ISSUE THREAD

**Issue by gzuidhof** [asg017/dataflow#4]
Hi Alex,

I'm building [Starboard Notebook](https://https://github.com/gzuidhof/starboard-notebook) (some demo noteboks on the frontpage [here](https://starboard.gg)), a notebook system for the browser. Much like dataflow it also features notebooks which are defined as plaintext files that are easy-to-git.

Different than this project however is that notebooks are entirely static: you don't need any server running at all. Your system would [embed the notebook](https://github.com/gzuidhof/starboard-wrap) in an iframe (and talk to it to for instance inject any secrets).

Starboard is built around dynamic importing of functionality: you can arbitrarily add new "cell types". For instance it supports Python in the browser, and recently I managed to hack together talking to a remote [Jupyter kernel](https://https://github.com/gzuidhof/starboard-jupyter). 

It would be really cool to support a cell type that is "Observable", I would love to work together on this. What it will provide Dataflow is a runtime that is sandboxed by default: a Starboard notebook runs entirely on a different origin in an iframe (editor included!), as well as a live notebook environment with code editor (as opposed to editing the source text files) - so pretty similar to Observable!

Now imagine interop with Python in the browser.. A workflow could be: "Input CSV file" -> "run some Pandas code in Python through Pyodide" -> "Visualize using Observable cell"... Super powerful stuff! 

I can ramble for a long time about why this would be great, but I'd love to hear your thoughts on perhaps adopting Starboard as the notebook "runtime" here. Happy to answer any questions as well of course! (happy to chat somewhere if its easier)

**Comment by asg017** [comment]
Hey @gzuidhof , this is a pretty cool idea, starboard seems awesome!

This should definitely be possible. Dataflow is just an opinionated wrapper + dev experience around the [unofficial-observablehq-compiler](https://github.com/asg017/unofficial-observablehq-compiler). With that compiler, you're able to compile "Observable javascript" into vanilla javascript (an ES module), or you can interpret "Observable javascript" live in the browser.

Dataflow does both of these things: with `dataflow run`, it creates a dev server that "inteprets" the Observable javascript you write live in the browser. But, with `dataflow export`, it compiles the Observable javascript into a flat ES module, which can be used later with the Observable runtime to run a notebook natively.

So, for Starboard, I'd imagine you could use `unofficial-observablehq-compiler` to convert Observable javascript into "plain" javascript, and run it in an iframe (and just refresh the entire iframe when code updates). This is similar to what I did with [atom-observable](https://github.com/asg017/atom-observable), which made the code pretty easy to implement, and didnt require a server or anything to run. You may not need to use Dataflow directly, since Dataflow sets up an entire devserver for it's own thing. Also, with Dataflow, when you update code, it surgically re-arranges the rendered notebook to optimize for not re-computing cells that don't need to be re-computed, but if you don't care, you can just refresh the entire page/notebook and be done with it much faster.


Another thing to consider is how you decide to render an Observable notebook in. Starboard itself. Since Observable code that looks like this:


```javascript
width = 400

viewof color = html`<input type=color value="#ff0000">`

d3 = require("d3");

chart = {
   d3.create("svg").selectAll(....)
}
```

Would render 4 elements like this:

![image](https://user-images.githubusercontent.com/15178711/114284420-9bd66d80-9a04-11eb-8f8b-ebb421e1ece1.png)


But starboard seems to only render the defined element/variable. But that kindof goes against Observable's "write your code in whatever order you like", but it's not *too* big of a deal to change around.

Another thing: I'm actively thinking about [changing unofficial-observablehq-compiler](https://github.com/asg017/unofficial-observablehq-compiler/issues/26) in the near future, so I'd follow that issue to get updates!

Anyway, tldr is that this is definitely possible, working with unofficial-observablehq-compiler alone would be more helpful that Dataflow specifically, and if you want to talk about this more directly with me, feel free to email me directly!

**Comment by gzuidhof** [comment]
Cool! Thank you for the reply :)

Thanks for explaining Dataflow more, it makes total sense. I suppose I misunderstood the goal of the project to be more in line of: a self hostable Observable editor (in which case Starboard would hopefully be able to save you a lot of work).

 I am not super familiar with Observable. I come from more of a Jupyter background, which is much less pure and not reactive (run cells in any order, have internal state, etc).
My intention for Starboard was to build a general base that can power both types of paradigm, and the browser happens to be built with vanilla JS built-in, so that's the glue/base of everything. Any other runtimes or languages such as Python through Pyodide/WASM or Observable/Pluto.jl can be plugged in on demand.

> But starboard seems to only render the defined element/variable. But that kindof goes against Observable's "write your code in whatever order you like", but it's not too big of a deal to change around.

You're right, it just renders the value you return (or something you print). But a cell type's implementation isn't bound to that: you just get a few HTML elements to mount on and anything is fair play from there. The Observable cell type should retain the "write your code in whatever order you like" behavior: that's what makes Observable so great! So for sure this cell will have to behave differently :). An existing example of such behavior is the CSS cell: it just applies on every change without a "run" button.

One thing I am a bit scared about is the runtime and exceptions. Observable's own editor has a crazy setup in which there is another iframe behind where you actually type the code, and the page "in front" has transparent bits to show the iframe. I wonder if the Observable runtime can ever get in such a terrible state that it crashes the entire thing. One way to find out I suppose :)

I'll see if I can get something going with your unofficial compiler, thank you again!

**Comment by gzuidhof** [comment]
![observable](https://user-images.githubusercontent.com/1039510/114287644-619ab980-9a60-11eb-8dea-8aeb1db674e3.gif)

It's getting there. Let me know if you would rather have me create an issue in the compiler project instead (I figured I would post here first so I don't pollute it there)

I am not entirely fluent in Observable so there may be stuff I'm misunderstanding. Right now I'm doing this: I create a global runtime and module, which is shared between all cells in my notebook. Then for every cell I create a single observer into a local HTML element.

The above gif starts with the following notebook:
```python
# %%--- [esm]
# properties:
#   run_on_load: true
# ---%%
import {registerObservable} from "http://localhost:8080/dist/starboard-observable.js";

registerObservable();

# %% [observable]
a = 1
# %% [observable]
b = "asdf"
```

Whenever a character is typed I define or redefine the cell with this logic:

```javascript
try {
    if (!this.observableCell) { // The first time the cell is run this will be undefined
        this.observableCell = await compile.cell(this.cell.textContent);
        this.observableCell.define(main, this.observer);
    } else {
        this.observableCell = await compile.cell(this.cell.textContent);
        this.observableCell.redefine(main)
    }
} catch (e) {
    console.error(e);
    this.errorElement.innerText = e.toString();
}
```

This naive approach has the following limitations:
* Cells can not be deleted. If I start of with a cell that states `a=5` there will still be that value after I delete the cell. Do I need to wait for [this PR](https://github.com/asg017/unofficial-observablehq-compiler/pull/23)? Anything I can do to push it over the finish line?
* I have no idea whether to "redefine" or "define" a value whenever a new character is typed.  
   As you start typing `abc = 123` it will define a variable `a` as that is the first character typed, and then it goes all wrong. Anonymous cells also don't work as they should always be "defined" instead of "redefined. Is there any way for me to know?
I guess I just need to know the name of the defined variable, or `null` in case of anonymous.

If you're curious about the code, [here](https://github.com/gzuidhof/starboard-observable) are the ~100 lines or so :).

**Comment by gzuidhof** [comment]
Thinking about it a bit more it doesn't make sense to run on every keystroke as even Observable has a "run" button. But the two questions still stand :)

**Comment by Kreijstal** [comment]
Have you read this? 

https://observablehq.com/@observablehq/how-observable-runs

Basically, there is a tree of dependencies, like on excel, and it has to know which cells depend on which other cells, so when a cell is updated, the parent cells that depend on that cell are updated as well.

**Comment by jcmkk3** [comment]
Since this thread is all about collaboration between projects, I wanted to draw your attention to another project that I saw start up recently and already seems to be picking up a lot of steam. One of the Jupyter maintainers (@jtpio) is working on a WASM/client side version of Jupyter called [jupyterlite](https://github.com/jtpio/jupyterlite). Maybe there's an opportunity there?
